feat(develop): Federated Analytics : Boxplot and five number summary core

msuzen opened the pull request:
* Federated boxplot with five number summary.

Commit:
feat: added boxplot provider and five number class

Providers for Federated analytics computing five numbers and boxplot

nfnt commented on components/kubeflow-pipeline/pipeline_components/analytics-client/flwr_analytics_client/fedbox.py:
```diff
@@ -0,0 +1,4 @@
+from fedhist import HistogramProvider
+
+class BoxPlotProvider(HistogramProvider):
+    pass
```
Please implement the `name` property for this provider, matching that of the server-side one, as it will be used to dispatch the server input to the right provider.

```suggestion
    @property
    def name(self) -> str:
        return "boxplot"
```

msuzen replied:
Good catch. This is updated.

nfnt commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fedbox.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class BoxPlotProvider(AnalyticsProvider):
+    def __init__(self, hrange:Tuple) -> None:
```
Nit: Let's remove the `hrange` parameter here, as it'll be set in `set_parameters`. This way we will have a defined flow of events that set these values.

msuzen replied:
Added this into `set_arguments`.

nfnt commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fivenum.py:
```diff
@@ -0,0 +1,68 @@
+
+
+@dataclass
+class fiveNumSummary:
+    """Five number summary."""
+
+    histogram: HistogramData
+    hmin: float
+    hmax: float
+    fivenum: Optional[fiveNum] = None
```
Let's use data classes to group similar parameters, not to group functionality related to five number summary, as that's already captured by having this package.
With this current approach, it's easy to overlook which of the parameters are input values and expected to set when creating an instance of this class and which values will be set when calling `compute`. Instead, let's have a single function `compute_fivesum` and a helper function `inverse_value_bin`. By having:

```python
def compute_fivesum(hmin: float, hmax: float, histogram: HistogramData) -> FiveNum
```

it is clear which values are input and what is being output by the function. As there's only 3 input parameters, it's fine to name them directly, though once we have more parameters, it makes sense to create a separate data class to capture all input values like all output which are captured with the `fiveNum` class.

msuzen replied:
Make sense. Updated.

nfnt commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fivenum.py:
```diff
@@ -0,0 +1,68 @@
+
+
+@dataclass
+class fiveNum:
```
Nit: Let's use _upper camel case_ here to follow Python convention for classes.

```suggestion
class FiveNum:
```

msuzen replied:
Updated.

nfnt reviewed (changes requested):
This is looking great, thanks for this great addition to our analytics toolset!
I have a few minor comments to ensure that everything is working as expected.

Commit:
feat: added boxplot provider and five number class

Addressed the PR comments. name property in boxplot client. Server provider arguments set mechanism.
FiveNum dataclass only and computation move into helper function. Class naming convention.

nfnt reviewed (approved):
LGTM!

GerardoGR commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fedbox.py:
```diff
@@ -0,0 +1,101 @@
+from .provider import AnalyticsProvider, scalar_to_numpy
+
+
+class BoxPlotProvider(AnalyticsProvider):
```
It looks like this provider does a similar job as the `HistogramProvider`. Consider inheriting from the `HistogramProvider` which will allow you to only override the `aggregate`, `result_metadata_json` and `name` methods

```suggestion
class BoxPlotProvider(HistogramProvider):
# ...
```

msuzen replied:
Good catch. Updated.

GerardoGR commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fedbox.py:
```diff
@@ -0,0 +1,101 @@
+            )
+        )
+
+    def aggregate(self) -> None:
+        client_input = self.client_input_data()
+        hmin = client_input['hmin']
+        hmax = client_input['hmax']
+        histogram=aggregate_histograms(self._client_data)
+        fn = compute_fivesum(hmin, hmax, histogram)
+        self._result = fn
```
Following the `HistogramProvider` inheritance consideration. This method could look like:

```suggestion
    def aggregate(self) -> None:
        super().aggregate()
        histogram = self._result

        client_input = self.client_input_data()
        self._result = compute_fivesum(client_input['hmin'], client_input['hmax'], histogram)
```

msuzen replied:
Makes sense. Updated.

Commit:
feat: added boxplot provider and five number class

Addressed the PR comments. BoxPlotProvider inheriting from HistogramProvider, having less
methods rewritten. Change the mechanism where aggregation uses super's aggregate method first.

GerardoGR commented on components/kubeflow-pipeline/pipeline_components/analytics-server/flwr_analytics_server/fedbox.py:
```diff
@@ -0,0 +1,60 @@
+from argparse import ArgumentParser, Namespace
+from typing import Dict, List, Tuple
+from .fivenum import compute_fivesum, FiveNum
+from .fedhist import aggregate_histograms, HistogramProvider, HistogramData
```
nit: There are a bunch of unused imported libraries, e.g the `aggregate_histograms` is not used. Please double check which libraries are not used and remove them.

GerardoGR reviewed (approved):
Other than my nit-comment. I think the PR looks good to go! Good job @msuzen !

Commit:
feat: added boxplot provider and five number class

Unused imports in fedbox is removed

Commit:
feat: added boxplot provider and five number class

Unused imports in fedbox is removed. Second round
